File: UnknownSizeArray.java
// TC : O(log n)
// SC :  log(1)

//We need to  keep moving right until we find number that is less than target
//Then use binary search algo
public class class UnknownSizeArray {
    public int search(ArrayReader reader, int target) {

        int low  = 0;
        int high = 10;
        int mid = 0;

        while(low <=high){
            mid = (low + high)/2;

            if(reader.get(mid) == target)
                return mid;
            // binary search
            if(reader.get(low) < target <= reader.get(high)){
                if(reader.get(mid) < target)
                    low = target +1;

                if(reader.get(mid) >= target){
                    high = mid -1;
                }


            }
            else{   // move to the right side with bigger range
                low = high +1;
                high = 2* high;
            }


        }
        return -1;

    }

File: UnknownSizeArray.py
# Time complexity : O(log(m) + log(k)), m is the pointer where high pointer ends and k is the range in which binary search is performed.
# Space complexity : O(1)

# The code ran on leetcode

def binary_search(target):
    l = 0; h = 1
    if reader.get(l) == target:
        return 0
    
    # Update low and high pointers such that the target lies between them
    while reader.get(h) < target:
        l = h; h = 2*h

    # Once the range is found, perform binary search in that range
    while l <= h:
        mid = l + (h-l)//2

        if reader.get(mid) == target:
            return mid
        # Search left portion of the array
        elif reader.get(mid) > target:
            h = mid - 1
        # Search right portion of the array
        else:
            l = mid - 1